### DISCUSSION TITLE
Problemi di comunicazione MQTT, Docker, zigbee2mqtt e Broker Mosquitto

### DISCUSSION THREAD

**Discussion by simonevisciano** [indomus/forum#3097] [GENERALE - Discussioni generiche]
Ciao ragazzi, da premeterre che non sono un esperto di questi servizi ma grazie a voi sono riuscito a mettere in piedi la mia domotica con Home Assistant (Core) su Mini-PC con tutto ciò che ne conviene:
- accesso in remoto con Cloudflare;
- sistema Alexa completo;
- MQTT;
- Zigbee;
- Docker Compose gestito su Portainer;
- ecc.

Quindi prima di tutto grazie e complimenti per il lavoro che svolgete, guide super dettagliate, news ecc, fate semplicemente paura.

Volevo raccontarvi/sfogare una mia esperienza in questo viaggio che magari può aiutare qualcuno per quanto riguarda proprio MQTT (non so se la sezione è giusta).

Sono stati due giorni intensi, dopo aver repullato lo stack che comprendeva HA, mosquitto, zigbee2mqtt e altro, ho cominciato ad avere problemi con il sistema MQTT, disconnessioni continue con l'alternarsi di periodi invece stabili, per farla breve non c'era un cavolo di pattern che mi facesse capire chi o cosa stesse causando instabilità nel sistema.
Tutto il problema alla fine riguardava tre container, ha, mosquitto e zb2mqtt, i log che ricevevo da loro erano sensa senso (per farmi capire il problema), mosquitto non dava errori o non inviava messaggi, gli altri due invece errori di disconnessioni, quando solo uno, quando entrambi, publish di messaggi che non avvenivano, a volte invece si verificavano duplicati (anche se QoS a 0), un incubo.

Dopo reset di configurazioni, modifiche di tutti i colori nel conf sia di mosquitto che di zb2mqrr, studio di documentazioni di tutti per cercare qualunque cosa che mi facesse capire qualcosa, arrivo alla fine alla conclusione che il problema era la sola comunicazione tra loro più che i settaggi. Sono arrivato a pensato perfino che fosse qualcosa a livello di router/modem, firewall o altro, pacchetti persi, blocchi di controllo di comunicazioni TCP, protezioni quindi che bloccano a volte si e a volte no i messaggi MQTT anche se in realtà sapevo che non aveva senso visto che la comunicazione dei tre container era in localhost, in sostanza il problema erano i network creati in docker, alcuni container erano sotto stesso network altri no, mi sembrava comunque strano visto che ripeto le comunicazione MQTT erano sempre sulla stessa macchina, non usciva nulla in pratica, non passava per router, switch o altro (frontend escluso), decido quindi di fare l'ennesima prova, stoppare tutti i container, eliminare tutti i network creati e riavviari i container, problema risolto, lo stack docker-compose con ha, mosquitto, zb2mqtt ecc si sono messi tutti sotto la stessa sotto-rete e ora se accendo una luce da alexa quindi "Alexa>Cloud AWS>HA>Broker>ZB2MQTT>Switch Zigbee", l'azione fisica viene eseguita in meno di mezzo secondo, se la faccio da HA direttamente (quindi tutto il locale) è instantanea.

Il mio consiglio è casomai di specificarlo questo concetto del network anche se riguarda il sistema Docker o di aggiungerlo agli esempi di config docker la voce network così da forzare la subnet di appartenza dei container.
Ora forse era ovvio il problema, non sono un sistemista o un esperto, fatto sta che non ho trovato nulla neanche su internet, neanche su forum inglesi, magari ho utilizzato le keyword sbagliate non so.
Comunque, evidentemente le comunicazioni tra sottoreti diverse di container diversi da questi problemi/latenze anche se nella mia ignoranza, mi sembra comunque strano, forse avrei risolto comunque specificando un gateway preciso? Non lo so, diciamo però che nel campo dell'informatica le vie di mezzo sono abbastanza rare, le cose di solito o funzionano o no. Qualcuno sa darmi delucidazioni per pura curiosità?

Scusate la lunghezza, spero che la mia esperienza posso essere di aiuto a qualcuno, non mollate mai ragazzi!

**Comment by snakuzzo** [comment]
Guarda, entrare troppo nel dettaglio delle network di docker, confonderebbe solo gli utenti, soprattutto quelli che sono alle prime armi e non sono così pratici. Online c'è la documentazione ufficiale e chiunque può andare a leggerla.
Nel tuo caso, se stai usando docker-compose, mi sembra strano che tu abbia avuto container su network separate.
Quando tiri su un' "applicazione" con docker compose, di default, se non viene specificato in maniera esplicita, docker crea una network con una subnet nella quale verranno inclusi tutti i servizi (container) di quel file yaml.
Se da quello che dici ti sei trovato in questa situazione "strana", è perché sicuramente ti sei perso qualcosa nel tempo...strada facendo.

**Reply by simonevisciano** [reply]
Assolutamente, qualcosa mi sono perso io strada facendo come dici, ne sono convinto, probabilmente quando ho fatto il trasferimento in portainer.

**Reply by snakuzzo** [reply]
sì, potrebbe essere.
Io comunque nel compose ho definito esplicitamente la network di docker, indicando anche la subnet e, tra container, la comunicazione la faccio con nome container, in modo che tra loro comunicheranno sempre, anche se la subnet cambia, e non sono dipendenti dalla network dell'host (salvo ovviamente i container in modalità host come Core...dove, anziché puntare l'IP LAN del server o la loopback, puoi recuperare l'IP del gateway di docker e puntare quello.

**Reply by simonevisciano** [reply]
Bello mi piace in effetti, molto più sicuro, vedrò di implementare la cosa.
Grazie del suggerimento.
